stl/inc/algorithm

```diff
@@ -2797,6 +2797,220 @@ namespace ranges {
     };
 
     _EXPORT_STD inline constexpr _Fold_right_last_fn fold_right_last{_Not_quite_object::_Construct_tag{}};
+
+    class _Find_last_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <forward_iterator _It, sentinel_for<_It> _Se, class _Ty, class _Pj = identity>
+            requires indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>
+        _NODISCARD constexpr subrange<_It> operator()(_It _First, _Se _Sent, const _Ty& _Value, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Sent);
+
+            auto [_UFirst, _ULast] = _Find_last_unchecked(
+                _Unwrap_iter<_Se>(_STD move(_First)), _Unwrap_sent<_It>(_STD move(_Sent)), _Value, _Pass_fn(_Proj));
+            _It _Last{};
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            _Seek_wrapped(_Last, _STD move(_ULast));
+            return {_STD move(_First), _STD move(_Last)};
+        }
+
+        template <forward_range _Rng, class _Ty, class _Pj = identity>
+            requires indirect_binary_predicate<ranges::equal_to, projected<iterator_t<_Rng>, _Pj>, const _Ty*>
+        _NODISCARD constexpr borrowed_subrange_t<_Rng> operator()(
+            _Rng&& _Range, const _Ty& _Value, _Pj _Proj = {}) const {
+            auto _First            = _RANGES begin(_Range);
+            auto [_UFirst, _ULast] = _Find_last_unchecked(
+                _Unwrap_range_iter<_Rng>(_STD move(_First)), _Uend(_Range), _Value, _Pass_fn(_Proj));
+            iterator_t<_Rng> _Last{};
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            _Seek_wrapped(_Last, _STD move(_ULast));
+            return {_STD move(_First), _STD move(_Last)};
+        }
+
+    private:
+        template <class _It, class _Se, class _Ty, class _Pj>
+        _NODISCARD static constexpr subrange<_It> _Find_last_unchecked(
+            _It _First, _Se _Last, const _Ty& _Value, _Pj _Proj) {
+            _STL_INTERNAL_STATIC_ASSERT(forward_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>);
+
+            if constexpr (_Bidi_common<_It, _Se>) {
+                for (auto _Result = _Last; _Result != _First;) {
+                    if (_STD invoke(_Proj, *--_Result) == _Value) {
+                        return {_STD move(_Result), _STD move(_Last)};
+                    }
+                }
+                return {_Last, _Last};
+            } else if constexpr (same_as<_It, _Se>) {
+                auto _Result = _Last;
+                for (; _First != _Last; ++_First) {
+                    if (_STD invoke(_Proj, *_First) == _Value) {
+                        _Result = _First;
+                    }
+                }
+                return {_STD move(_Result), _STD move(_Last)};
+            } else {

```

Comment:
There are 4 possible combinations of bidi and common. We've handled "bidi and common" and "forward and common" cases above. The remaining cases, "bidi non-common" and "forward non-common" are possible here. The implementation below (in this `else` block) is the only possible implementation for "forward non-common", of course.

However, for the "bidi non-common" case, this is using the forward-only algorithm, which invokes the projection and predicate on every single element. It would be possible to find the last iterator, and then use the bidi algorithm. For bidi-only iterators, this would pay N traversals only, followed by O(distance from result to last) backwards-traversals/projections/predicates. While that isn't a 100% guaranteed win, it is likely to be better, especially for expensive predicates and when we're fortunate to find the result near the last. And if the arguments are random and sized, we can jump to the end instantly. (This scenario could occur when the iterator/sentinel pair is `vector::iterator/const_iterator`, for example.)

Fortunately, there's a helper to make this easier, `_Get_final_iterator_unwrapped`. The callers (iterator/sentinel and range) can have `if constexpr` for bidi-or-stronger iterators, and use `_Get_final_iterator_unwrapped` before calling the unchecked helper. Then the `_Bidi_common` case in the unchecked helper would absorb this scenario, so a fourth case would *not* need to be added.

Please see existing calls to `_Get_final_iterator_unwrapped`, or ask @CaseyCarter for assistance as he has volunteered to help. :smile_cat: